### ISSUE TITLE
JOSS review #1

### ISSUE THREAD

**Issue by marouenbg** [VU-BEAM-Lab/GENRE#1]
Hi @Christopher-Khan ,

I will post my reviews here as I go through the software
https://github.com/openjournals/joss-reviews/issues/2644

**Comment by marouenbg** [comment]
GENRE is a GPU implementation of elastic-net that allows to handle several regression problems in parallel. It is provided as a CUDA implementation with an interface to MATLAB through MEX that allows a larger user base to access the software. The package is well documented, it is implemented in a scripting language and a programming language.

Here are my initial comments:

- You mentioned that GENRE was twice faster in GPU than CPU. Can you add a table with the benchmark results?

- Consider adding community guidelines or a contribution guide, consider moving the `contribution` section in README to a `contribute.md`

- The packages needs a manual (a test function with a toy example) or an automatic testing component (Travis/ Github actions / Appveyor).

In the manuscript:

- The multiple references format is the following [@author1:2001; @author2:2001]` -> "(Author1 et al., 2001; Author2 et al., 2002)

- The full reference of MATLAB is (The MathWorks, Inc., Natick, MA, USA) 

- In the statement of need section, please specify the acronym: Aperture Domain Model Image
REconstruction (ADMIRE)

- In 'However, a GPU implementation of the
algorithm was developed' I think you need to specify that you're talking about GENRE here

- extend the acronyms fMRI, CT, PET in words

**Comment by marouenbg** [comment]
The README file is quite complete, I had a couple of comments about the format.

- In the README `setup` section, please make the installation instructions as ordered bullet points with a specific header for Windows and Linux rather than a paragraph. You can add a troubleshooting step for each bullet point

- same for `model data format` section, consider having bullet points for the variables and the files

- transformation flag values in `user-defined parameters` also could be in bullet points

- I think that README and tutorial are quite informative, consider transforming them into a MATLAB livescript if you think this could be useful.

Going back to my first comment about benchmarking results, how do you think GENRE would perform in comparison to a multi-CPU LASSO implementation? I am interested in the effect of clock speed in CPUs vs GPUs.

**Comment by Christopher-Khan** [comment]
Hi @marouenbg! Thank you for taking the time to review my submission and for your helpful comments. I really appreciate it. I have addressed your comments about the manuscript and will continue to address the remaining comments. I was a little slow to respond because I've been attending a virtual conference this week. I just wanted to ask about two of the above comments.

1. For the comment that talks about GENRE being two times faster than a CPU, is this referring to the section in my paper where I say that the GPU implementation of ADMIRE was two orders of magnitude faster than the CPU implementation? I can post the benchmarks, but I was able to optimize the GPU elastic-net even more in that specific case. For example, groups of model fits could use the same model matrix in that case, which allowed me to achieve improved coalesced memory access and to better utilize GPU memory bandwidth. Essentially, I had developed the GPU elastic-net for our ADMIRE project and then decided to make it a generalized software package (GENRE). As a result of this generalization, I just assumed each model fit uses a unique model matrix. Therefore, did you still want me to include the ADMIRE-specific benchmarks. I was thinking it might be better to include a benchmark comparing GENRE to a serial C implementation because it can still provide significant speedup. In terms of how it would compare to a multi-CPU implementation, I haven't benchmarked that (I haven't used OpenMP or MPI to create a multi-CPU implementation), but I think it would be possible to get similar speedup depending on how many CPU cores are available and the resources that are available for each core. However, a user could take the GPU kernels from GENRE and optimize them even more for their specific application like I did for ADMIRE, which is why I tried to extensively comment the code :). For example, in the case of ADMIRE, optimizing it even more made the difference of going from one order of magnitude speedup to two orders of magnitude when compared to a serial CPU implementation. In a case like this, it will be more difficult for a multi-CPU implementation to achieve similar performance. Did you have a specific multi-CPU package in mind? I ask because many of the papers I've seen usually focus on using multiple CPUs for just one model fit and not multiple in parallel.

2. In regards to the toy model, I do have a tutorial in the README, which involves generating toy data and processing it using GENRE. However, I can add more if it is not sufficient. 

Thank you again for reviewing this submission.

**Comment by marouenbg** [comment]
Hi @Christopher-Khan ,
Thanks for the feedback, I had fun reviewing this quality contribution.

Regarding question 1, yes if you can post the benchmark where GENRE is twice as fast in comparison to CPU and also mention the ADMIRE specific contribution, that would support the central claim in this paper. 
About multi-CPU implementation, I think your answer above could be added in the discussion section and that would be enough in my opinion, there is no need to add more benchmarks.

Regarding question 2, I meant that you can use the toy model to implement a test functionality using [TRAVIS-CI](https://travis-ci.org/) for example. If you're not familiar with continuous integration, you can write a series of small tests (which can be based on your tutorial) and use the `assert` function to determine that software and hardware are properly configured. You can call the function `testGENRE()` and it could be a quick way for users to tell if everything is working properly.

**Comment by marouenbg** [comment]
Continuing with the review:

- I run the tutorial on a K80 machine, AWS p2.xlarge. I think `GENRE.m` should be a function that can be called as `GENRE()` or `GENRE_Tutorial()` with arguments. Maybe also `data_organizer.m`.

- What is the rationale for this computation https://github.com/VU-BEAM-Lab/GENRE/blob/master/GENRE_Code/GENRE.m#L111

**Comment by marouenbg** [comment]
- Is `GENRE_GPU_double_precision.cu` a standalone, i.e., can it be installed through a compiler without MATLAB? If so, I suggest adding the installation instructions and a minimal working example.

- I went through the cuda file and I think you did a flawless implementation with a very clear and easy-to-follow code, congratulations! I liked the part where you optimized the memory access through local and shared to further optimize the run time. It could be worthwhile to discuss this feature of the implementation (memory types and speed) in the manuscript.

- Memory is often the limiting resource on GPU, do you expect applications or model size that would require more than the capacity of current GPUs? Do you think that an implementation in "streams" like many machine learning packages have, would be address this type of issues?

- Final thought, what are the advantages of your implementation over coding it directly in MATALB using the parallel toolobx and the `gpuArray` routine?

These are my final thoughts. I think the implementation is very clean, with very clear code and docs. You went the extra mile to implement the software in cuda and used mex to make it callable from MATLAB. This is an excellent piece of software.

**Comment by Christopher-Khan** [comment]
Hi @marouenbg. Thank you for the additional comments about the code! It's great to have a fellow CUDA programmer approve of the CUDA implementation :). Also, I'm really glad to hear the code was easy to follow. I had spent a good amount of time trying to clean it up and comment it.

I've addressed some of your additional comments. I've added bullet points to the relevant sections, and I have added a Contributing section to the README.md file. My contribution guidelines are relatively simple, which is why I just added a section in the README.md instead of making a separate contribute.md file. I will work on addressing the rest of the comments in the next few days. I will add the benchmark in the manuscript. In terms of the test functionality, I actually haven't used TRAVIS-CI before, but I'll take a look at it. If I don't use that, I will just use your other suggestion and write a testGENRE() function that creates a small dataset and fits it like in my tutorial. I guess I can have the output just be a plot of the original y signal along with the y signal that the model fit produces. The user will be able to see if the model fit has actually been performed or not based off of this.

In terms of making GENRE a function instead of a script, I was actually hoping to get feedback on this from the review process, so thank you for bringing it up. I had just kept it as a script in case if MATLAB users wanted to access the workspace variables once the script finishes. So in your opinion, do you think a function format is more user-friendly and simpler? If so, do you think I should just include the model coefficients as the only function output? Also, I can make the data_organizer.m script be a function too, but it will just end up using arguments that you would pass into the GENRE function, which is why I had just kept it as a script.

In terms of the rationale for the computation that you had pointed out, in that section, I see if shared memory can be used. If so, I compute the maximum number of threads that can be used within one block. I then find the closest power of 2 that is below this number. Essentially, I'm just making sure the block size is a multiple of the warp size, which is 32 (some of the older GPUs have a warp size of 16). You're probably familiar with this already due to your experience using CUDA, but using a multiple of the warp size helps ensure that resources aren't being wasted. For example, suppose I have 32 model fits to perform, but I set 31 as the number of threads to use within the block. This will waste resources because two blocks will be needed. The first block will have a warp where 31 threads are doing work, and one thread is just masked out. The other block will just have 1 thread doing work with the others masked out. This is wasteful because you have to issue instructions to two warps in this case. Moreover, depending on the problem, you might be wasting memory bandwidth in this scenario if all 32 model fits need access to values that are stored in contiguous memory locations. This is because if you had used a block size of 32 instead of 31, then only one warp would be executed, and one coalesced memory transaction could be used to service all of the threads whenever a memory access occurs.

In terms of GENRE_GPU_double_precision.cu, it can't be used as a standalone as is, but it can be with modification. For example, if you just wanted to use the C/CUDA code without MATLAB, you would just need to get rid of the components in the file that are associated with the MEX-interface. These are mainly the parts where data is being passed back and forth with MATLAB (towards the beginning and end of the file). You would then need to determine how you want to access the data variables once the MEX-interface is removed. For example, you can just read from and write to binary files. The nice thing is that the lines of code that actually call the CUDA kernels and the CUDA kernels themselves would remain the same. In terms of compiling, you should just be able to compile it with nvcc. Now, if you want to use another high level language, then you would need to replace the MEX-interface with whatever interface the language uses. Languages such as R, Python, and Julia all have some way to call CUDA kernels. I didn't include any specific instructions because there are a variety of ways to potentially use it without MATLAB.

In terms of memory being a limit, I completely agree. Given the large spectrum of problem domains that this software can potentially be applied to, I think there will be situations where the GPU memory capacity will not be enough. I agree with your suggestion of stream processing as being a potential solution, especially if the size of each individual model fit is very large. With this approach though, you would also have to consider the performance penalty from having to perform multiple data transfers between host and device. In addition, in the case where you want to perform an extremely large number of model fits and can't fit all of the fits onto one GPU, then distributing the computations across multiple GPUs could also be another approach.

In terms of the advantages of a custom CUDA implementation vs MATLAB gpuArrays, my general opinion is that writing custom CUDA code has more potential to be tuned, especially if it is being tuned to a specific problem. For example, in this application, using shared memory is an example optimization that was used, and I'm not sure if the gpuArray approach would be able to perform the same optimization in this case. Therefore, due to this, I would expect the custom approach to have more performance benefits. I believe that the MATLAB gpuArrays will be most beneficial to people who do not necessarily want to dive deep into CUDA but would still like to achieve some additional performance benefits. The gpuArrays only support certain functions, so this may or may not be sufficient for a given application. Moreover, another benefit of using custom CUDA code for this application is that it makes it easier to use with languages other than MATLAB and as a standalone because you already have all of the source code. However, MATLAB has some tools to generate source code as well, but I haven't really used them. Adding on, I personally like custom CUDA code because it allows you to have more control with what's going on under the hood in terms of implementation and resource management.

This post was really long haha, but your comments and feedback have been very helpful in improving this repository!

**Comment by marouenbg** [comment]
Thank you for addressing all these questions in great detail.

> So in your opinion, do you think a function format is more user-friendly and simpler?

yes, I think it could help to have a `GENRE()` function.

**Comment by Christopher-Khan** [comment]
Hi @marouenbg. I have converted GENRE.m from a script to a function. In addition, I have added a test_GENRE.m function, and I have updated the Setup section in the README.md file to explain it. Is this sufficient for the testing component from your perspective?

Thank you for your time.

**Comment by Christopher-Khan** [comment]
In addition to my above comment, I have also added a test_GENRE_shared_memory.m function. It does the same thing as the test_GENRE.m function, but it uses smaller model matrix sizes to test for the case where shared memory is used in addition to global memory for the model fits. I have described it in the Setup section as well.

**Comment by marouenbg** [comment]
This is perfect @Christopher-Khan. I will finish checking the last tick box in my review. The cache of your latex code `objective_function` in the README file does not seem to work.

**Comment by marouenbg** [comment]
One last thing, if you're going to keep the speed up claims in the paper, they have to be backed up by benchmark data of GPU vs CPU. Also add the CPU and GPU hardware specifications that you're testing on.

**Comment by Christopher-Khan** [comment]
Hi @marouenbg. Sorry for the late response. I have added the ADMIRE benchmark to the paper. One thing that I updated is that we had originally used glmnet for the CPU implementation, but I recently wrote a custom C implementation of cyclic coordinate descent that is faster. Therefore, we use this for the CPU implementation of ADMIRE now, and I have updated the paper to reflect this. In addition, I liked your idea of adding an extra section to the paper demonstrating the performance comparison when using shared memory and when not using shared memory. I can also add a benchmark comparing GENRE to glmnet in this section. I will try and do this by the end of this week.

Thank you for your time.
